src/stores/Schedules.ts

```diff
@@ -0,0 +1,488 @@
+        this.addUndoState();
+        for (const scheduleIndex of scheduleIndices) {
+            const customEvents = this.schedules[scheduleIndex].customEvents;
+            const index = customEvents.findIndex((customEvent) => customEvent.customEventID === customEventId);
+            if (index !== undefined) {
+                customEvents.splice(index, 1);
+            }
+        }
+    }
+
+    /**
+     * Change color of a custom event
+     * @param customEventId
+     * @param newColor
+     */
+    changeCustomEventColor(customEventId: number, newColor: string) {
+        this.addUndoState();
+        const customEvent = this.getExistingCustomEvent(customEventId);
+        if (customEvent) {
+            customEvent.color = newColor;
+        }
+    }
+
+    /**
+     * Replaces properties of custom event with ones from editedCustomEvent and moves the custom event to newIndices.
+     * Edits the custom event object itself so all references are edited.
+     * @param editedCustomEvent
+     * @param newIndices
+     */
+    editCustomEvent(editedCustomEvent: RepeatingCustomEvent, newIndices: number[]) {
+        this.addUndoState();
+        const customEvent = this.getExistingCustomEvent(editedCustomEvent.customEventID);
+        if (customEvent === undefined) {
+            this.addCustomEvent(editedCustomEvent, newIndices);
+            return;
+        }
+
+        // Modify the original custom event so all references are updated as well
+        Object.assign(customEvent, editedCustomEvent);
+
+        const currentIndices = this.getIndexesOfCustomEvent(editedCustomEvent.customEventID);
+        // Equivalent to currentIndices set minus newIndices
+        const indicesToDelete = currentIndices.filter((index) => !newIndices.includes(index));
+        this.deleteCustomEvent(customEvent.customEventID, indicesToDelete);
+
+        // Equivalent to newIndices set minus currentIndices
+        const indicesToAdd = newIndices.filter((index) => !currentIndices.includes(index));
+        this.addCustomEvent(customEvent, indicesToAdd);
+    }
+
+    /**
+     * Checks if a schedule contains the custom event ID
+     * @param customEventId
+     * @param scheduleIndex
+     */
+    doesCustomEventExistInSchedule(customEventId: number, scheduleIndex: number) {
+        for (const customEvent of this.schedules[scheduleIndex].customEvents) {
+            if (customEvent.customEventID === customEventId) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    // --- Other methods ---
+    /**
+     * Appends a copy of the current state to previous states to revert to.
+     * Previous states are capped to 50 to prevent memory usage from going infinite.
+     */
+    addUndoState() {
+        const clonedSchedules = JSON.parse(JSON.stringify(this.schedules)) as Schedule[]; // Create deep copy of Schedules object
```

Comment:
```suggestion
        const clonedSchedules = structuredClone(this.schedules);
```

Reply:
Damn I swear when I looked up deep copying I didn't see this.

Reply:
Bro same, I learned about it from  [YouTube shorts](https://youtube.com/shorts/XK0V0E3bA-M?feature=share) lol. I swear StackOverflow never mentioned anything about this wizardry